lib/SILOptimizer/Mandatory/TFDeabstraction.cpp

```diff
@@ -1786,10 +1817,97 @@ void TFDeabstraction::cleanupDeadInstructions() {
     removeArrayValueIfPossible(inst);
 }
 
+/// If the specified type conforms to the TensorProtocol protocol, return the
+/// Scalar type for it.  Otherwise return a null type.
+static Type conformsToTensorProtocol(Type ty, ModuleDecl *module) {
+  auto nominal = ty->getAnyNominal();
+
+  auto &ctx = ty->getASTContext();
+  auto tensorProto = ctx.getProtocol(KnownProtocolKind::TensorProtocol);
+  if (!tensorProto || !nominal)
+    return Type();
+
+  SmallVector<ProtocolConformance *, 2> conformances;
+  nominal->lookupConformance(/*unused module*/ nullptr, tensorProto,
+                             conformances);
+  if (conformances.size() != 1)
+    return Type();
+
+  auto scalarMembers =
+      nominal->lookupDirect(DeclName(ctx.getIdentifier("Scalar")));
+  if (scalarMembers.size() != 1)
+    return Type();
+  if (auto member = dyn_cast<TypeDecl>(scalarMembers[0]))
+    return ty->getTypeOfMember(module, member,
+                               member->getDeclaredInterfaceType());
+  return Type();
+}
+
+/// Given something that conforms to the TensorProtocol protocol, extract the
+/// 'handle' out of it.
+static SILValue getTensorProtocolHandleMember(SILValue v, SILLocation loc,
+                                              SILBuilder &B) {
+  // If we already have a TensorHandle, just use it.
+  if (isTensorHandle(v->getType()))
+    return v;
+
+  auto module = B.getFunction().getModule().getSwiftModule();
 
-// TODO: move this from TFUtilities when deabstraction is done.
-SILValue getTensorProtocolHandleMember(SILValue v, SILLocation loc,
-                                       SILBuilder &B);
+  auto vType = v->getType().getASTType();
+  if (!vType->getStructOrBoundGenericStruct() ||
+      !conformsToTensorProtocol(vType, module))
+    return SILValue();
+
+  // If this value is just a struct wrapper around a TensorHandle, use the
+  // input of it.  In the case of Tensor2D, we have multiple levels of struct
+  // wrapper.
+  while (1) {
+    auto si = dyn_cast<StructInst>(v);
+    if (!si)
+      break;
+
+    if (si->getNumOperands() != 1)
+      break;
+
+    auto operand = si->getOperand(0);
+    // If we found the TensorHandle itself, then we win - return it.
+    if (isTensorHandle(operand->getType()))
+      return operand;
+
+    // If we found a wrapper around another TensorProtocol, dig deeper.
+    if (!conformsToTensorProtocol(operand->getType().getASTType(), module))
+      break;
+
+    v = operand;
+  }
+
+  // TODO(clattner): it would be more correct to generate a call to an accessor
+  // to get the handle out, but for now, we know we're always dealing with types
+  // that store the field by-value so we can dig it out in with an easier
+  // approach.  This handles structs of TensorHandle (like Tensor) and structs
+  // of structs of TensorHandle (like Tensor2D).
```

Comment:
`Tensor2D` no longer exists. Since this patch is a code cleanup, could you refresh this comment as well? Thanks!

Reply:
Done, thanks!